src/mon/OSDMonitor.cc

```diff
@@ -10191,8 +10192,20 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
 	  rule_name = poolstr;
 	}
       }
+      cmd_getval(g_ceph_context, cmdmap, "expected_num_objects",
+                 expected_num_objects, int64_t(0));
     } else {
       //NOTE:for replicated pool,cmd_map will put rule_name to erasure_code_profile field
+      //     and put expected_num_objects to ruleset field
+      if (rule_name != "") {
+        string interr;
+        expected_num_objects = strict_strtoll(rule_name.c_str(), 10, &interr);
+        if (interr.length()) {
+          ss << "error parsing integer value '" << rule_name << "': " << interr;
+          err = -EINVAL;
+          goto reply;
+        }
+      }
```

Comment:
This doesn't work.  For example,
<pre>
2018-01-12 03:30:18.208 7f08b542c700  1 -- 172.21.15.182:0/4125772450 --> 172.21.15.182:6789/0 -- mon_command({"format": "json", "rule": "by-host-smithi182", "pool_type": "replicated", "prefix": "osd pool create", "pg_num": "128", "erasure_code_profile": "by-host-smithi182", "pool": "by-host-smithi182"} v 0) v1 -- 
0x55b9e857ff80 con 0
2018-01-12 03:30:18.212 7f08bbf29700  1 -- 172.21.15.182:0/f5ea4aa2 <== mon.0 172.21.15.182:6789/0 11 ==== mon_command_ack([{"format": "json", "rule": "by-host-smithi182", "pool_type": "replicated", "prefix": "osd pool create", "pg_num": "128", "erasure_code_profile": "by-host-smithi182", "pool": "by-host-smithi182
"}]=ffffffea error parsing integer value 'by-host-smithi182': Expected option value to be integer, got 'by-host-smithi182' v10) v1 ==== 151+0+0 (883dec20 0 0) 0x55b9e857ff80 con 0x55b9e845a700
</pre>

Reply:
@liewegas 
I can reproduce your problem througth 'ceph -c ceph.conf osd pool create by-host-smithi182  128 replicated by-host-smithi182 **by-host-smithi182**'.
The expected format of this command is 'ceph osd pool create {pool-name} {pg-num} [{pgp-num}] [replicated] [crush-rule-name] [**expected-num-objects**]'.
So the last 'by-host-smithi182' is parsed as 'expected-num-objects'. 
All behaviours are expected.

Reply:
In this case the command isnt coming from the CLI, it's the localpool mgr module, and it's sending well-formed JSON.  We need to make OSDMonitor handle a well-formed command.

(The weird positional CLI parsing is super-lame, but it is what we have right now.  We should avoid breaking the JSON in order to fix the CLI...)